edge/pkg/devicetwin/dmiserver/server.go

```diff
@@ -0,0 +1,263 @@
+	}
+
+	for _, twin := range in.ReportedDevice.Twins {
+		propertyType, ok := twin.Reported.Metadata[PropertyType]
+		if !ok {
+			errLog := fmt.Sprintf("fail to get propertyType for property %s of device %s", twin.PropertyName, in.DeviceName)
+			klog.Errorf(errLog)
+			return nil, fmt.Errorf(errLog)
+		}
+		msg, err := CreateMessageTwinUpdate(twin.PropertyName, propertyType, twin.Reported.Value)
+		if err != nil {
+			klog.Errorf("fail to create message data for property %s of device %s with err: %v", twin.PropertyName, in.DeviceName, err)
+			return nil, err
+		}
+		handleDeviceTwin(in.DeviceName, msg)
+	}
+
+	return &pb.ReportDeviceStatusResponse{}, nil
+}
+
+func handleDeviceTwin(deviceName string, payload []byte) {
+	topic := dtcommon.DeviceETPrefix + deviceName + dtcommon.TwinETUpdateSuffix
+	target := modules.TwinGroup
+	resource := base64.URLEncoding.EncodeToString([]byte(topic))
+	// routing key will be $hw.<project_id>.events.user.bus.response.cluster.<cluster_id>.node.<node_id>.<base64_topic>
+	message := beehiveModel.NewMessage("").BuildRouter(modules.BusGroup, modules.UserGroup,
+		resource, messagepkg.OperationResponse).FillBody(string(payload))
+
+	beehiveContext.SendToGroup(target, *message)
+}
+
+// CreateMessageTwinUpdate create twin update message.
+func CreateMessageTwinUpdate(name string, valueType string, value string) (msg []byte, err error) {
+	var updateMsg DeviceTwinUpdate
+
+	updateMsg.BaseMessage.Timestamp = getTimestamp()
+	updateMsg.Twin = map[string]*types.MsgTwin{}
+	updateMsg.Twin[name] = &types.MsgTwin{}
+	updateMsg.Twin[name].Actual = &types.TwinValue{Value: &value}
+	updateMsg.Twin[name].Metadata = &types.TypeMetadata{Type: valueType}
+
+	msg, err = json.Marshal(updateMsg)
+	return
+}
+
+func initSock(sockPath string) error {
+	klog.Infof("init uds socket: %s", sockPath)
+	_, err := os.Stat(sockPath)
+	if err == nil {
+		err = os.Remove(sockPath)
+		if err != nil {
+			return err
+		}
+		return nil
+	} else if os.IsNotExist(err) {
+		return nil
+	} else {
+		return fmt.Errorf("fail to stat uds socket path")
+	}
+}
+
+func StartDMIServer(mapperList map[string]*pb.MapperInfo, deviceList map[string]*v1alpha2.Device, deviceModelList map[string]*v1alpha2.DeviceModel) {
+	err := initSock(SockPath)
+	if err != nil {
+		klog.Fatalf("failed to remove uds socket with err: %v", err)
+		return
+	}
+
+	lis, err := net.Listen(deviceconst.UnixNetworkType, SockPath)
+	if err != nil {
+		klog.Errorf("failed to start DMI Server with err: %v", err)
+		return
+	}
+
+	limiter := rate.NewLimiter(rate.Every(Limit*time.Millisecond), Burst)
+
+	s := grpc.NewServer()
+	pb.RegisterDeviceManagerServiceServer(s, &server{
+		limiter:         limiter,
+		mapperList:      mapperList,
+		deviceList:      deviceList,
+		deviceModelList: deviceModelList,
+	})
+	reflection.Register(s)
+
+	if err := s.Serve(lis); err != nil {
+		klog.Errorf("failed to start DMI Server with err: %v", err)
```

Comment:
If err !=nil then return, otherwise, all the `success` messages and `error` messages will be logged.

Reply:
done